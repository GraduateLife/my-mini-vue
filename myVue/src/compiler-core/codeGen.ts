import { AstNodeTypes } from "./astNodeTypes";
import {
	CREATE_ELEMENT_NODE,
	HelperMap,
	TO_DISPLAY_STRING,
} from "./transformHelpers";

export function generate(ast) {
	const context = createCodeGenContext();
	const { helpers } = ast;
	const { concat } = context;

	const functionName = "render";
	const functionArgs = ["_ctx", "_cache"].join(", ");
	if (helpers.length > 0) {
		importFunctionHelpers(ast, context);
	}

	concat("return function ");

	concat(`${functionName}(${functionArgs}){`);

	concat("return ");

	processCodeGenNode(ast.codeGenNode, context);

	concat("}");
	// console.log("code be like", context.code);
	return {
		code: context.code,
	};
}

function processCodeGenNode(node, context) {
	// console.log("genNode is like", node);
	switch (node.type) {
		case AstNodeTypes.TEXT:
			processTextNode(node, context);
			break;
		case AstNodeTypes.INTERPOLATION:
			processInterpolationNode(node, context);
			break;
		case AstNodeTypes.SIMPLE_EXPRESSION:
			processSimpleExp(node, context);
			break;
		case AstNodeTypes.ELEMENT:
			processElementNode(node, context);
			break;
		case AstNodeTypes.MIXED_EXPRESSION:
			processMixedExpression(node, context);
			break;
		default:
		// throw new Error("impossible node type");
	}
}

function processTextNode(node, context) {
	const stuff = node.content;
	context.concat(`'${stuff}'`);
}
function processInterpolationNode(node, context) {
	const { concat, getHelper } = context;
	concat(getHelper(TO_DISPLAY_STRING));
	concat("(");
	processCodeGenNode(node.content, context);
	concat(")");
}
function processSimpleExp(node, context) {
	context.concat(`${node.content}`);
}

function processElementNode(node, context) {
	// console.log("multi", node.branches);
	const { concat, getHelper } = context;
	const { tag, attributes, branches } = node;
	concat(getHelper(CREATE_ELEMENT_NODE));
	concat("(");
	processElementArray(
		mustNotReturnUndefined([tag, attributes, branches]),
		context
	);
	// concat(`'${tag}'`);

	// concat(`, ${attributes}, `);
	// const insideElement = node.branches[0]; //DONE:not good
	// processCodeGenNode(branches, context); //since we know there is only one node
	// concat(`'this is', +_toDisplayString(_ctx.message)`);
	// for (let i = 0; i < node.branches.length; i++) {
	// 	processCodeGenNode(node.branches[i], context);
	// }
	concat(")");
}
const mustNotReturnUndefined = (someArray) => {
	return someArray.map((item) => item || "null");
};
function processElementArray(elementArr, context) {
	const { concat } = context;
	for (let i = 0; i < elementArr.length; i++) {
		if (typeof elementArr[i] === "string") {
			concat(elementArr[i]);
		} else {
			processCodeGenNode(elementArr[i], context);
		}
		if (i < elementArr.length - 1) {
			concat(", ");
		}
	}
}

function processMixedExpression(node, context) {
	for (let i = 0; i < node.branches.length; i++) {
		const { concat } = context;
		const currNode = node.branches[i];
		if (typeof currNode !== "string") {
			processCodeGenNode(currNode, context);
		} else {
			concat(currNode);
		}
	}
}

function createCodeGenContext() {
	const context = {
		code: "",
		concat(something) {
			context.code += something;
		},
		getHelper(helperName) {
			return `_${HelperMap[helperName]}`;
		},
	};
	return context;
}

function importFunctionHelpers(ast, context) {
	const { concat } = context;
	const Vue = "Vue";

	const { helpers } = ast;
	const setHelperAlias = (helperSymbol) => {
		return `${HelperMap[helperSymbol]}: _${HelperMap[helperSymbol]}`;
	};
	concat(`const { ${helpers.map(setHelperAlias)} } = ${Vue}`);
	concat("\n");
}
