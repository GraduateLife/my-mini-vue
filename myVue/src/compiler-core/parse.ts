import { AstNodeTypes } from "./astNodeTypes";

enum ElementTagsIdentifiers {
	START, //< of <>
	END, //</ of </>
}

export function baseParse(input: string) {
	const context = createParseContext(input);
	return createRoot(parseBranches(context, []));
}
function parseBranches(context, recentElements) {
	// console.log(
	// 	"start parsing context branch,current source is,,",
	// 	context.source
	// );
	const nodes: Array<any> = [];
	let node;
	while (!shouldTerminateParse(context, recentElements)) {
		// console.log("current element stack", recentElements);
		if (/^\{\{/.test(context.source)) {
			// {{something...
			node = parseInterpolation(context);
		} else if (/^\<\w/.test(context.source)) {
			// <something...
			node = parseElement(context, recentElements);
		} else {
			node = parseText(context);
		}

		nodes.push(node);
	}

	return nodes;
}
function parseInterpolation(context) {
	const openBrace = "{{";
	const closeBrace = "}}";
	const closeBraceIndex = context.source.indexOf(closeBrace, openBrace.length);
	proceedByLength(context, openBrace.length); //proceed to right by 2(  {{  )
	//console.log("do we have problem when we at {{?>>>", context.source);
	const parsedInterpolationContent = processWordContent(
		context,
		closeBraceIndex - openBrace.length
	).trim();
	//console.log(
	//"do we have problem when we are inside interpolation?>>>",
	//parsedInterpolationContent,
	//",,should be parsed expression,",
	//context.source,
	//",,is current source"
	//);
	proceedByLength(context, closeBrace.length); //move to the end of }}
	//console.log(
	//"do we have problem when we at }}?>>>",
	//context.source,
	//",,is current source"
	//);
	// console.log("ip ends", context.source);
	return {
		type: AstNodeTypes.INTERPOLATION,
		content: {
			type: AstNodeTypes.SIMPLE_EXPRESSION,
			content: parsedInterpolationContent,
		},
	};
}

function parseElement(context, recentElements) {
	const element: any = parseElementTags(context, ElementTagsIdentifiers.START);
	recentElements.push(element);
	// console.log(
	// 	"current element stack after parsing element tag start",
	// 	recentElements
	// );
	element.branches = parseBranches(context, recentElements);
	recentElements.pop();
	// console.log(
	// 	"current element stack after parsing element tag",
	// 	recentElements
	// );
	if (hasCorrespondingCloseTag(context, element)) {
		// console.log("start parsing close tag");
		parseElementTags(context, ElementTagsIdentifiers.END);
	} else {
		throw new Error("lost close tag");
	}

	return element;
}

function hasCorrespondingCloseTag(someCxt, someEle) {
	return (
		someCxt.source.slice(2, someEle.tag.length + 2) === someEle.tag &&
		someCxt.source.startsWith("</")
	);
}
function parseElementTags(
	context,
	elementTagsIdentifier: ElementTagsIdentifiers
) {
	const matches = /^\<\/?(\w+)\>/.exec(context.source);
	// //console.log("tag matches results", matches);
	const parsedElementTagName = matches![1];
	// console.log(
	// 	"do we have problems when we parsing tag names?>>>",
	// 	parsedElementTagName,
	// 	",,should be parsed tag"
	// );
	proceedByLength(context, matches![0].length);
	// console.log("current source", context.source);

	if (elementTagsIdentifier === ElementTagsIdentifiers.END) {
		return;
	}
	return {
		type: AstNodeTypes.ELEMENT,
		tag: parsedElementTagName,
	};
}

function parseText(context: any) {
	let textEndIndex = context.source.length;
	const textEndSymbols = ["{{", "<"];

	for (let i = 0; i < textEndSymbols.length; i++) {
		const temp = context.source.indexOf(textEndSymbols[i]);
		// console.log("qqqqqqqqqqqqq", temp);
		if (temp !== -1 && temp < textEndIndex) {
			textEndIndex = temp;
		}
	}

	const parsedText = processWordContent(context, textEndIndex);
	//console.log(
	//"parsed text is,,",
	//parsedText,
	//"current source is,,",
	//context.source
	//);
	return {
		type: AstNodeTypes.TEXT,
		content: parsedText,
	};
}
function processWordContent(context, someLength) {
	const parsedText = context.source.slice(0, someLength);
	proceedByLength(context, someLength);
	return parsedText;
}

function shouldTerminateParse(context, recentElements) {
	if (context.source.startsWith("</")) {
		for (let i = recentElements.length - 1; i >= 0; i--) {
			let lastElement = recentElements[i];
			// console.log("aaaaaaaaa", lastElement);
			if (hasCorrespondingCloseTag(context, lastElement)) {
				// console.log("out");
				return true;
			}
		}
	}

	if (context.source.length === 0) {
		return true;
	}
	return false;
}
//to proceed string pointer
function proceedByLength(someCxt, length: number) {
	someCxt.source = someCxt.source.slice(length);
}

function createRoot(branches) {
	return {
		type: AstNodeTypes.ROOT,
		branches,
	};
}
function createParseContext(input: string) {
	return {
		source: input,
	};
}
