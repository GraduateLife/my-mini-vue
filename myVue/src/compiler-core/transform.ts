import { AstNodeTypes } from "./astNodeTypes";
import { TO_DISPLAY_STRING } from "./transformHelpers";

export function transform(rootNode, options = {}) {
	const context = createTransformContext(rootNode, options);

	dfs(rootNode, context);
	createRootCodeGenNode(rootNode);
	rootNode.helpers = [...context.helpers.keys()];
}
function createRootCodeGenNode(rootNode) {
	const firstChild = rootNode.branches[0];
	if (firstChild.type === AstNodeTypes.ELEMENT) {
		rootNode.codeGenNode = firstChild.codeGenNode;
	} else {
		rootNode.codeGenNode = firstChild;
	}
}

function dfs(node, context) {
	// if (node.type === AstNodeTypes.TEXT) {
	// 	node.content = "this is transformed message";
	// }
	//execute plugins before go into it
	const { nodeTransforms } = context;
	const onLeaveExecutionLIFO: any = [];
	for (let i = 0; i <= nodeTransforms.length - 1; i++) {
		const _exec = nodeTransforms[i];
		const deferred = _exec(node, context);
		if (deferred) {
			onLeaveExecutionLIFO.push(deferred);
		}
	}
	switch (node.type) {
		case AstNodeTypes.INTERPOLATION:
			context.addHelper(TO_DISPLAY_STRING);
			break;
		case AstNodeTypes.ROOT:
		case AstNodeTypes.ELEMENT:
			passThoughBranch(node, context);
			break;
		default:
	}

	let i = onLeaveExecutionLIFO.length;
	while (i--) {
		onLeaveExecutionLIFO[i]();
	}
}
function passThoughBranch(node, context) {
	// console.log("branch is like", node.branches);
	const branches = node.branches;
	for (let i = 0; i < branches.length; i++) {
		const nextNode = branches[i];
		dfs(nextNode, context);
	}
}
//transform configurations
function createTransformContext(rootNode, options) {
	const context = {
		root: rootNode,
		nodeTransforms: options.nodeTransforms || [],
		helpers: new Map(),
		addHelper(helperName) {
			context.helpers.set(helperName, null);
		},
	};
	return context;
}
