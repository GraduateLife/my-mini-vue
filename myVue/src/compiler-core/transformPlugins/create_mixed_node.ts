import { AstNodeTypes } from "../astNodeTypes";
const isLiteralNode = (someNode) => {
	return (
		someNode.type === AstNodeTypes.INTERPOLATION ||
		someNode.type === AstNodeTypes.TEXT
	);
};
//node1,    node2,      node3,
//A         B           C             <<<before
//A+B       C                         <<<after

export function create_mixed_node(node) {
	// console.log("node is like", node);

	let mixedExpContainer;
	if (node.type === AstNodeTypes.ELEMENT) {
		return () => {
			for (let i = 0; i < node.branches.length; i++) {
				const currNode = node.branches[i];
				if (isLiteralNode(currNode)) {
					// console.log("currrrrrrr", currNode);
					for (let j = 1 + i; j < node.branches.length; j++) {
						const nextNode = node.branches[j];
						if (isLiteralNode(nextNode)) {
							// console.log("nexxxxxxxxxxt", nextNode);
							if (!mixedExpContainer) {
								//init
								mixedExpContainer = {
									type: AstNodeTypes.MIXED_EXPRESSION,
									branches: [currNode],
								};
								node.branches[i] = mixedExpContainer; //replace current node to mixed-exp container and ready to push new members
							}
							mixedExpContainer.branches.push(" + "); //push +
							mixedExpContainer.branches.push(nextNode); // push its neighbor
							node.branches.splice(j, 1); //remove this node(the neighbour) from current branch
							j--; //rollback pointer
						} else {
							mixedExpContainer = null;
							break;
						}
					}
				}
			}
		};
	}
	// console.log("container", mixedExpContainer);
}
