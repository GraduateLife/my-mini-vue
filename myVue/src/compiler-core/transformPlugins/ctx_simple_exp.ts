import { AstNodeTypes } from "../astNodeTypes";

export function ctx_simple_exp(node) {
	if (node.type === AstNodeTypes.INTERPOLATION) {
		node.content = add_ctx(node.content);
		// node.content.content = "_ctx." + node.content.content;
	}
}
function add_ctx(insideIPcontent: any) {
	insideIPcontent.content = `_ctx.${insideIPcontent.content}`;
	return insideIPcontent;
}
