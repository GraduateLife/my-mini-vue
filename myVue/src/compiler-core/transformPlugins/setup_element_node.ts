import { AstNodeTypes, createElementCodeGenNode } from "../astNodeTypes";

export function setup_element_node(node, transformCxt) {
	if (node.type === AstNodeTypes.ELEMENT) {
		return () => {
			//setup element codegenNode
			const elementTag = `'${node.tag}'`;
			let elementAttrs;
			let elementBranches = node.branches[0]; //setup the codegenNode need to be handled

			node.codeGenNode = createElementCodeGenNode(
				transformCxt,
				elementTag,
				elementAttrs,
				elementBranches
			);
		};
	}
}
