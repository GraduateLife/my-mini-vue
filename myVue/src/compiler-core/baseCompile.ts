import { transform } from "./transform";
import { generate } from "./codeGen";
import { baseParse } from "./parse";
import { create_mixed_node } from "./transformPlugins/create_mixed_node";
import { ctx_simple_exp } from "./transformPlugins/ctx_simple_exp";
import { setup_element_node } from "./transformPlugins/setup_element_node";

export function baseCompile(template: string) {
	const ast: any = baseParse(template);
	transform(ast, {
		nodeTransforms: [ctx_simple_exp, setup_element_node, create_mixed_node],
	});

	return generate(ast).code;
}
