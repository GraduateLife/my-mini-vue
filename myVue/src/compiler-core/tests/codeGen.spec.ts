import { transform } from "../transform";
import { generate } from "../codeGen";
import { baseParse } from "../parse";
import { ctx_simple_exp } from "../transformPlugins/ctx_simple_exp";
import { setup_element_node } from "../transformPlugins/setup_element_node";
import { create_mixed_node } from "../transformPlugins/create_mixed_node";
import { baseCompile } from "../baseCompile";

describe("code generator", () => {
	it("creates code for strings", () => {
		const ast = baseParse("good morning");
		transform(ast);
		const { code } = generate(ast);

		expect(code).toMatchSnapshot();
	});
	it("creates code for interpolations", () => {
		const ast = baseParse("{{message}}");
		transform(ast, {
			nodeTransforms: [ctx_simple_exp],
		});
		const { code } = generate(ast);

		expect(code).toMatchSnapshot();
	});
	it("create code for elements", () => {
		const ast = baseParse("<div></div>");
		transform(ast, {
			nodeTransforms: [setup_element_node],
		});
		const { code } = generate(ast);

		expect(code).toMatchSnapshot();
	});

	it("create code for multiple types", () => {
		const ast: any = baseParse("<div>this is {{message}}</div>");
		transform(ast, {
			//to specify, in this case, we let transformer to pass element first child only to the generator, hence we have to execute mixed-exp plugin
			//first to convert all its innertext into one node (mixed type)
			// nodeTransforms: [ctx_simple_exp, create_mixed_node, setup_element_node], //<<<you will suddenly find simple-exp plugin no longer works
			//#region what happened inside?
			/*it is due to the arrangement of ast nodes, every node will try to execute all plugins
			first in : element type (<div>),tells generator to process its first child only,(actually it kills all other children),
			but before that, executes mixed-node plugin, to aspect the type of its children
			second in: text type('this is'), triggered mixed-exp plugin, convert all the innertext into one node, so it's safe in codegen process
			third in? No we don't have the third node any more, since interpolation type is merged into mixed-exp type(removed from branch), 
			hence simple-exp plugin doesn't have a chance to be executed,
			What shall we do?
			we want simple-exp plugin to be executed as soon as possible, if we can't do that, we defer the execution timing of other plugins

			*/
			//#endregion
			nodeTransforms: [ctx_simple_exp, setup_element_node, create_mixed_node],
		});
		// console.log("ast is like", ast.codeGenNode.branches);
		const { code } = generate(ast);

		expect(code).toMatchSnapshot();
	});
	it("tiny up", () => {
		const piece = baseCompile("<div>this is {{message}}</div>");
		expect(piece).toMatchSnapshot();
	});
});
