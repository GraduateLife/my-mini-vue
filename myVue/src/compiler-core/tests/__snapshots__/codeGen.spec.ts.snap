// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code generator create code for elements 1`] = `
"const { createElementNode: _createElementNode } = Vue
return function render(_ctx, _cache){return _createElementNode('div', null, null)}"
`;

exports[`code generator create code for multiple types 1`] = `
"const { toDisplayString: _toDisplayString,createElementNode: _createElementNode } = Vue
return function render(_ctx, _cache){return _createElementNode('div', null, 'this is ' + _toDisplayString(_ctx.message))}"
`;

exports[`code generator creates code for interpolations 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`;

exports[`code generator creates code for strings 1`] = `"return function render(_ctx, _cache){return 'good morning'}"`;



exports[`code generator tiny up 1`] = `
"const { toDisplayString: _toDisplayString,createElementNode: _createElementNode } = Vue
return function render(_ctx, _cache){return _createElementNode('div', null, 'this is ' + _toDisplayString(_ctx.message))}"
`;
